Better diagnostics for multi-line string literals

beccadax opened the pull request:
Improves the error messages emitted when a multiline string literal has a whitespace error. Incorrect indentation uses one of these three messages:

* "string literal content is insufficiently indented"
* "string literal content is indented with a tab where a space is expected"
* "string literal content is indented with a space where a tab is expected"

All three of these errors also include two fix-its:

* "change indentation to match last line"
* "change last line's indentation to match this line"

Finally, missing newlines after the opening delimiter or before the closing delimiter are indicated with:

* "multi-line string literal content must begin on a new line"
* "multi-line string literal content must not be on the delimiter's line"

Both of these also include fix-its to add newlines.

I was not able to build a toolchain and test this in Xcode, so I'm not totally sure this will give the experience we hope for. But I think this is the right direction.

Resolves [SR-4701](https://bugs.swift.org/browse/SR-4701).

@milseman @johnno1962

beccadax committed:
Add tests for improved multiline string errors

beccadax committed:
Add more detailed indentation errors

beccadax committed:
Add fix-it to adjust current line’s indentation

beccadax committed:
Extract helper functions for string processing

beccadax committed:
Get CharSourceRange of indentation

beccadax committed:
Add suggestion to change indentation of string

beccadax committed:
Fix firstNotOf test

Also correct some mistakes in tests

beccadax committed:
Rearrange comments to improve clarity

beccadax committed:
Improve “no newline at beginning” error message

Including a new fix-it to add the newline.

beccadax committed:
Reword “no newline before literal end” error

beccadax committed:
Shorten lines over 80 columns wide

beccadax committed:
Remove now-unused generic error message

xwu commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -129,12 +129,19 @@ ERROR(lex_invalid_unicode_scalar,none,
+ERROR(lex_unexpected_space_in_string_indent,none,
+      "string literal content is indented with a space where a tab is expected", ())
+NOTE(note_change_current_line_indentation,none,
+      "change indentation to match last line", ())
```
How would you feel about: "change _this line's_ indentation to match last line"--not just for symmetry, but so that the message doesn't only mention the last line when the key thing is that we're offering to change the current line?

brentdax replied:
That works for me.

milseman replied:
Is there a more specific term than "last line" here? At first I thought it meant the previous line, but now I think you're referring to the line on which the closing delimiter is.

jrose-apple replied:
I agree with Michael, but I can't figure out how to phrase it non-awkwardly. "change this line's indentation to match that of the closing delimiter".

xwu commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -129,12 +129,19 @@ ERROR(lex_invalid_unicode_scalar,none,
-ERROR(lex_ambiguous_string_indent,none,
-      "invalid mix of multi-line string literal indentation", ())
-
+      "multi-line string literal content must not be on the delimiter's line", ())
```
Suggestion: "multi-line string literal closing delimiter must begin on a new line"

brentdax replied:
I might tweak this to mention indentation, but yeah, I think that'll work.

tkremenek commented:
@swift-ci clean test

milseman commented:
@jrose-apple Jordan, can you help bikeshed the content of the diagnostic messages? I don't have a strong intuition here, but the new errors feel different in style.

milseman commented:
I'll try in the meantime ;-)

milseman commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -129,12 +129,19 @@ ERROR(lex_invalid_unicode_scalar,none,
-
+      "multi-line string literal content must not be on the delimiter's line", ())
+ERROR(lex_insufficient_string_indent,none,
+      "string literal content is insufficiently indented", ())
```
hmm, maybe "insufficient indentation in multi-line string literal"?

jrose-apple replied:
Another option: "multi-line string literal must have consistent indentation"

milseman commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -129,12 +129,19 @@ ERROR(lex_invalid_unicode_scalar,none,
+ERROR(lex_insufficient_string_indent,none,
+      "string literal content is insufficiently indented", ())
+ERROR(lex_unexpected_tab_in_string_indent,none,
+      "string literal content is indented with a tab where a space is expected", ())
```
maybe "unexpected tab in multi-line string literal indentation; expected a space"

jrose-apple replied:
I actually like Brent's version better, although it might work better as a note on the previous error. Usually a single space doesn't slip in where a single tab is expected; instead you get multiple spaces for one tab. I assume there's a fix-it for the entire range of indentation but that could destroy any indentation the user actually intended, so it might even make sense to put it on a note instead of the main error.

milseman replied:
Yeah, it reads much better as a note.

milseman commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -129,12 +129,19 @@ ERROR(lex_invalid_unicode_scalar,none,
+ERROR(lex_unexpected_tab_in_string_indent,none,
+      "string literal content is indented with a tab where a space is expected", ())
+ERROR(lex_unexpected_space_in_string_indent,none,
+      "string literal content is indented with a space where a tab is expected", ())
```
maybe "unexpected space in multi-line string literal indentation; expected a tab"

milseman commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -129,12 +129,19 @@ ERROR(lex_invalid_unicode_scalar,none,
+      "change indentation to match last line", ())
+NOTE(note_change_last_line_indentation,none,
+      "change last line's indentation to match this line", ())
 
```
similarly WRT "last line"

xwu replied:
I have a suggestion: "change indentation of closing delimiter to match this line" and "change indentation of this line to match closing delimiter".

jrose-apple replied:
I don't think we should provide this fix-it at all. If this isn't the first line in the literal than other lines have already been shown to have proper indentation; if it *is* the first line you'll get a cascade of errors.

brentdax replied:
@jrose-apple Perhaps that fix-it should instead be a note pointing to the corresponding character in the delimiter indentation.

milseman reviewed (commented):
This is looking awesome, thanks Brent!

I have some minor bike shedding on diagnostic messages to try to bring them more in line with the rest of Swift error messages. But, I don't have a ton of experience in this matter, so please take them with a grain of salt.

beccadax committed:
Incorporate error message bikeshedding

beccadax committed:
Simplify return from getMultilineTrailingIndent

We no longer need the range, just the start location.

brentdax commented:
Okay, second try. This changes the errors to use more notes—for instance:

```
test/Parse/multiline_errors.swift:46:1: error: line does not start with same indentation as closing delimiter of multi-line string literal
        Pi
^
test/Parse/multiline_errors.swift:46:3: note: unexpected tab in indentation here
        Pi
  ^^^^^^
test/Parse/multiline_errors.swift:47:3: note: should match space in same position before closing delimiter
    """ // expected-error@-1{{line does not start with same indentation as closing delimiter of multi-line string literal}}
  ^
test/Parse/multiline_errors.swift:46:1: note: change indentation of this line to match closing delimiter
        Pi
^~~~~~~~
         
```

I'm still running the full test suite on this, but it passes all the multi-line tests.

xwu commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -131,17 +131,17 @@ ERROR(lex_unicode_escape_braces,none,
-      "string literal content is indented with a space where a tab is expected", ())
+      "multi-line string literal closing delimiter must begin on a new line", ())
+ERROR(lex_inconsistent_string_indent,none,
+      "line does not start with same indentation as closing delimiter of multi-line string literal", ())
```
Hmm, I think the note idea is great, but I feel like there are so many notes now. Can I make a few suggestions?

* This error message doesn't need to be so long; the fix-it explains exactly what to do already--this can just be the original "string literal content is insufficiently indented".

xwu commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -131,17 +131,17 @@ ERROR(lex_unicode_escape_braces,none,
+ERROR(lex_inconsistent_string_indent,none,
+      "line does not start with same indentation as closing delimiter of multi-line string literal", ())
+NOTE(note_insufficient_string_indent,none,
+      "unexpectedly found non-indentation character here", ())
```
* Is this necessary? If the line is insufficiently indented, then it should be pretty clear that the first non-indentation character is unexpected.

xwu commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -131,17 +131,17 @@ ERROR(lex_unicode_escape_braces,none,
+NOTE(note_insufficient_string_indent,none,
+      "unexpectedly found non-indentation character here", ())
+NOTE(note_unexpected_char_in_string_indent,none,
+      "unexpected %select{space|tab}0 in indentation here", (bool))
```
* Nit: doesn't need "in indentation"

xwu commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -131,17 +131,17 @@ ERROR(lex_unicode_escape_braces,none,
+NOTE(note_unexpected_char_in_string_indent,none,
+      "unexpected %select{space|tab}0 in indentation here", (bool))
+NOTE(note_matching_indentation_char_here,none,
+      "should match %select{space|tab}0 in same position before closing delimiter", (bool))
```
* Does this have to be a separate note? I feel like this and the above are one thought: "unexpected space instead of tab", etc. Surely, the user doesn't need to be pointed to a tab to know what it is?

brentdax replied:
Firstly, it's helpful to flag both locations because either one could be mistaken. But the main reason I do this is to illustrate the connection between the delimiter indentation and the line's indentation. I think this is important for people who are new to the multiline string feature and don't understand how its indentation behavior works yet.

It'd be nice to have a single note which pointed to both locations and said "these have to match", but our diagnostics don't allow that, so that leaves us with two separate notes. One of them could be merged with the error message, but both need to exist.

We could, perhaps, do this, though:

```
test/Parse/multiline_errors.swift:46:3: error: character does not match indentation before closing delimiter of multi-line string literal
        Pi
  ^^^^^^
test/Parse/multiline_errors.swift:47:3: note: tab there doesn't match corresponding space here
    """ // expected-error@-1{{line does not start with same indentation as closing delimiter of multi-line string literal}}
  ^
test/Parse/multiline_errors.swift:46:1: note: change indentation of this line to match closing delimiter
        Pi
^~~~~~~~
         
```

That would carry the same pieces of information in fewer lines

xwu replied:
That's a fair point about flagging both locations.

My thinking here is that, for better or worse, Swift error messages tend to be pretty terse. On the one hand, it's nice that you're making this message more helpful. On the other, I worry that the first reaction when a user triggers this error is, "Gah, what have I done? All these messages!"

Keep in mind that a user could have pasted 100 lines and indented only the closing delimiter inconsistently. Unless I'm mistaken, in that scenario, your design could generate 100 errors and 300 notes for a single misplaced tab! My instinct is that this will actually give people a distinctly negative impression of multiline literals, even though you're only trying to make them easier to use.

If we must have _three_ separate messages for each poorly indented line, then at least we can make each message more bite-sized. How do you feel about something like this:

* "error: unexpected indentation in multi-line string literal"
* "note: change indentation of this line to match closing delimiter"
* "note: indentation of preceding lines must start with these characters"--_emit this only once per literal_

jrose-apple replied:
Let's see. I think there are a few separate interesting cases here:

- Every line is consistent except the last one.
- Most lines are consistent with the last one, but a few are not.
- Everything is terrible and we can't tell what the indentation ought to be.

In the first case, the delimiter's clearly the one that's wrong, and it would be nice if we got one error and a note to fix it.

In the second case, we'd like to fix the offending line(s). I agree with Brent that it makes sense to show the final line as well, at least once, and I stand by my earlier comment that simply replacing all leading indentation may not be the correct fix and therefore should be on a note. The thing I might add to that is that if *consecutive* lines have the *same wrong indentation,* it might be nice™ to fix them all at once. But that's kind of a bonus.

The third case is just bad. Here, we could take a hint from Clang and just stop emitting errors after maybe the sixth mismatch. "note: suppressing any remaining indentation errors for this string literal"

None of this is a comment on the *text* of the diagnostic. I think Xiaodi's generally right that the notes can be a bit terser given the context of the error. The error should stand on its own but doesn't have to include the recommended fix.

xwu replied:
Excellent analysis, and great point about suppressing messages past a certain point. I would only quibble with the suggestion that these messages should all be errors.

The more I think about it, the more it seems to me that each (uninterpolated) multiline string literal should at most give rise to one error for indentation. After all, in every case the problem (well, the syntactic problem) is curable by deleting all whitespace preceding the closing delimiter. Of course, that may not give you the _intended_ string, and we wouldn't necessarily want to have that as a fix-it for people to mash on, but I think it does demonstrate that this is conceptually one error (with perhaps multiple places to be fixed, appropriate for notes) and not multiple.

jrose-apple replied:
Counterpoint: https://bugs.swift.org/browse/SR-4715. When the user makes a mistake, they want to fix it completely without having to recompile. That at least means one error with multiple notes.

(I really don't trust a fix-it for the whole thing because of content indentation:

```
func foo() {
  let defs = """
    (define (cdar x)
      (cdr (car x)) )
    (define (cadr x)
<TAB>  (car (cdr x)) )
    """
  print(defs)
}
```

)

xwu replied:
Agree on the one fell swoop; I'm advocating for one error + multiple notes.

brentdax replied:
The problem with one error + multiple notes is that, when you're using Xcode, the notes don't show up in the editor. So the Xcode experience would be that the entire literal is flagged as having indentation errors and there's a magic fix-it that tries to correct all of them, but no easy way to see which lines are erroneous.

Maybe we should split the difference: Merge together errors in adjacent lines, considering the common whitespace prefix[1] to be the indentation to remove. So if you wrote this (where the delimiter is indented with a tab, but the other lines have spaces):

```swift
let xml = """
    <html>
        <p>Hello, world!</p>
    </html>
	"""
```

You'd get something like:

```
example.swift:2:1: error: space in next 3 lines of multi-line string literal does not match tab indenting closing delimiter
    <html>
^
example.swift:3:1: note: also here
        <p>Hello, world!</p>
^
example.swift:4:1: note: also here
    </html>
^
example.swift:5:1: note: should match character here
        """
^^^^^^^^
example.swift:2:1 note: change indentation in these lines
    <html>
^~~~
        
        <p>Hello, world!</p>
^~~~
         
    </html>
^~~~~
        
```

([1] "Common whitespace prefix" is perhaps too simple an algorithm, since it would not detect when parts of a string were pasted from two different sources which were incompatible with the ending delimiter in different ways, but we could always switch to something that's more clever later on.)

jrose-apple replied:
I wouldn't bother with the "also here" notes when you've already said "next three lines" and will have a multi-line fix-it, but other than that this basically looks fine. I don't know if calling out "space" and "tab" specifically makes sense, though; it makes it sound like there's a one-for-one mismatch, and again you usually find multiple spaces to one tab. I would just say the indentation is mismatched and leave it at that.

xwu replied:
@brentdax Yeah, much improved. Agree with Jordan that "also here" and space vs. tab distinctions are probably unnecessary.

```
example.swift:2:1: error: unexpected indentation in multi-line string literal (3 lines)
    <html>
^^^^
example.swift:5:1: note: should match characters here
        """
^^^^^^^^
example.swift:2:1 note: change indentation of these lines
    <html>
^~~~
        
        <p>Hello, world!</p>
^~~~
         
    </html>
^~~~

```

beccadax committed:
Change indentation errors (again)

We now emit one error for each group of contiguous lines which have the same indentation mistake (i.e. the same common prefix with the expected indentation). That error mentions the number of lines and replaces only whatever indentation is common among those lines.

In cases where the entire literal is indented consistently with itself, but not with the closing delimiter, this should give you a single error covering the entire literal, with a single fix-it that consistently reindents the entire literal. In more complex cases, it still keeps the number of error messages manageable.

brentdax commented:
This latest set of revisions rebases on top of a more recent master, and also redesigns the errors again.

In a simple test case with a single error, you'll get something like this:

```
test/Parse/multiline_errors.swift:10:3: error: unexpected end of indentation of line in multi-line string literal
  Mu
  ^
test/Parse/multiline_errors.swift:11:3: note: should match space here
    """ // expected-error@-1{{unexpected end of indentation of line in multi-line string literal}}
  ^
test/Parse/multiline_errors.swift:10:3: note: change indentation of this line to match closing delimiter
  Mu
  ^
    
```

But it actually groups together consecutive lines with the error at the same location, so a single error message may cover several lines. In a [more complex case](https://github.com/apple/swift/pull/9148/commits/4e3915812d13288fc099c87377c00ff34715c4e7#diff-26a3c1073888782711f2c87657ecfb9cR87) with multiple segments, indented in mutually incorrect ways, you end up with something like:

```
test/Parse/multiline_errors.swift:88:1: error: unexpected end of indentation of next 2 lines in multi-line string literal
Zero A
^
test/Parse/multiline_errors.swift:100:1: note: should match tab here
                """   // expected-error@-12{{unexpected end of indentation of next 2 lines in multi-line string literal}}
^^^^^^^^
test/Parse/multiline_errors.swift:88:1: note: change indentation of these lines to match closing delimiter
Zero A
^
test/Parse/multiline_errors.swift:90:2: error: unexpected end of indentation of next 2 lines in multi-line string literal
        One A
        ^
test/Parse/multiline_errors.swift:100:2: note: should match tab here
                """   // expected-error@-12{{unexpected end of indentation of next 2 lines in multi-line string literal}}
        ^^^^^^^^
test/Parse/multiline_errors.swift:90:2: note: change indentation of these lines to match closing delimiter
        One A
        ^
test/Parse/multiline_errors.swift:92:1: error: unexpected space in indentation of next 4 lines in multi-line string literal
  Two A
^
test/Parse/multiline_errors.swift:100:1: note: should match tab here
                """   // expected-error@-12{{unexpected end of indentation of next 2 lines in multi-line string literal}}
^^^^^^^^
test/Parse/multiline_errors.swift:92:1: note: change indentation of these lines to match closing delimiter
  Two A
^
```

(The command line doesn't show multi-line fix-its, but this should emit one fix-it for each affected line, affecting only the indentation that's present in all of the lines, all controlled by a single note. Actually, I'd like someone to check that I'm correctly testing for that, because I have some doubts.)

jrose-apple commented:
Hm, yeah, we don't currently have a good way to specify a line number for fix-its right now. I think that's okay, although if you want to invent and implement a syntax for it (`line:col-line:col=`?) that could work too.

xwu commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -129,12 +129,17 @@ ERROR(lex_invalid_unicode_scalar,none,
-
+      "multi-line string literal closing delimiter must begin on a new line", ())
+ERROR(lex_inconsistent_string_indent,none,
+      "unexpected %select{space in|tab in|end of}2 indentation of "
```
I like this version quite a bit. My only suggestion here would be: `%select{unexpected space in|unexpected tab in|insufficient}2`. IMO, it's just much more easily interpreted when we call it "insufficient indentation of line" than "unexpected end of indentation of line." Looks good though.

xwu commented on include/swift/AST/DiagnosticsParse.def:
```diff
@@ -129,12 +129,17 @@ ERROR(lex_invalid_unicode_scalar,none,
+      "unexpected %select{space in|tab in|end of}2 indentation of "
+      "%select{line|next %1 lines}0 in multi-line string literal", 
+      (bool, unsigned, unsigned))
+NOTE(note_indentation_should_match_here,none,
```
Nit: for consistency, I think these would be named "lex_indentation_should_match_here" and "lex_change_indentation" below.

beccadax committed:
Improve phrasing of “end of indentation” message

beccadax committed:
Tweak diagnostic names and add assertions

beccadax committed:
Update tests

brentdax commented:
I've adopted @xwu's suggested "insufficient" message and renamed all three of those diagnostics to begin with `lex_multiline_string_indent_`. I think this is ready to go.

milseman commented:
This is looking great to me! 👍

milseman commented:
@swift-ci please test

milseman commented:
@swift-ci please smoke test Linux Platform

milseman commented:
@brentdax Linux fails with:
```
/home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/lib/Parse/Lexer.cpp:1467:30: error: use of undeclared identifier 'SIZE_T_MAX'
   size_t lastMistakeOffset = SIZE_T_MAX;
```

My number-fu is weak, but perhaps you want:
```
  uint64_t lastMistakeOffset = ~0ULL;
```

xwu commented:
I _think_ this will do:

```c++
#include <limits.h>
// ...
size_t lastMistakeOffset = std::numeric_limits<size_t>::max()
```

beccadax committed:
Use preferred C++ numeric size feature

SIZE_T_MAX isn’t available on Linux.

brentdax commented:
@milseman This last push should fix that (and matches some existing uses in the SIL optimizer, so it should be portable). Thanks to @xwu for pointing the way to `numeric_limits`.

milseman commented:
@swift-ci please smoke test Linux Platform

milseman commented:
@swift-ci please smoke test OS X Platform

milseman commented:
@brentdax I'm going to do a squash-and-merge. Do you have a preferred commit message?

brentdax commented:
@milseman I suppose something like this?

[Parse] Improve multi-line string literal errors

Adds descriptive errors and fix-its for multi-line string literal indentation and newline errors.
